clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java

```diff
@@ -95,13 +96,33 @@ public Builder setTargetTimes(List<ListOffsetsTopic> topics) {
 
         @Override
         public ListOffsetsRequest build(short version) {
+            for (ListOffsetsTopic topic : data.topics()) {
+                for (ListOffsetsPartition partition : topic.partitions()) {
+                    checkVersion(version, topic.name(), partition);
```

Comment:
The failed case `ListOffsetsRequestTest#testResponseDefaultOffsetAndLeaderEpochForAllVersions` can be reproduced on my laptop. Could you take a look? Thank you.

Reply:
Thank you, I will take a look!

Reply:
@frankvicky Could you please add a helper to build invalid request in order to run test?

Reply:
Hi @FrankYang0529 and @chia7712 
I have reviewed the code. Since this error is due to a version check in the build method, I will change `assertEqual` to `assertThrows` when the test tries to build an invalid request to align with the current behavior.

Reply:
@frankvicky What we want to protect is the request from clients (admin and consumer). So maybe we should revise `fromConsumer`[0] rather than `build`.

WDYT?

[0] https://github.com/apache/kafka/blob/21dd5cd421139d6134d738d6e7555f0db6e1f622/clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java#L65

Reply:
The benefit is that we can leverage the API_VERSION check of network layer. The disadvantage is that it requires the API_VERSION, and fortunately that is existent in consumer and admin.

Reply:
Hi @chia7712
I have reviewed the code. Since `forConsumer` is a static method, if we aim to perform a version check in this method, we will need to add a new argument. I will try to implement it.